src/plugins/expressions/public/interpreter_provider.ts

```diff
@@ -111,11 +117,29 @@ export function interpreterProvider(config: InterpreterConfig): ExpressionInterp
   async function invokeFunction(
     fnDef: AnyExpressionFunction,
     context: any,
-    args: Record<string, unknown>
+    args: Record<string, unknown>,
+    canCache = true
   ): Promise<any> {
     // Check function input.
     const acceptableContext = cast(context, fnDef.context ? fnDef.context.types : undefined);
-    const fnOutput = await fnDef.fn(acceptableContext, args, handlers);
+
+    let fnOutput;
+    // Caching: check if we already have results of the function
+    if (canCache && !fnDef.noCache) {
+      const hash = calculateObjectHash([
+        fnDef,
+        acceptableContext,
+        args,
+        handlers.getInitialContext(),
+      ]);
+      fnOutput = functionCache.get(hash);
+      if (!fnOutput) {
```

Comment:
Possible 0, "", false edge case? 
I'd use `functionCache.has(hash)` to check if there is a cache hit